nbdime/merging/decisions.py

```diff
@@ -0,0 +1,462 @@
+            action="either",
+            local_diff=local_diff,
+            remote_diff=remote_diff,
+            )
+
+    def conflict(self, path, local_diff, remote_diff):
+        assert local_diff and remote_diff
+        assert local_diff != remote_diff
+        action = "base"
+        self.add_decision(
+            path=path,
+            conflict=True,
+            action=action,
+            local_diff=local_diff,
+            remote_diff=remote_diff,
+            )
+
+
+def ensure_common_path(path, diffs):
+    """Resolves common paths in a list of diffs.
+
+    If a local and a remote diff both patch a key "a", this will return the
+    common path ("a",), and the inner diffs of the patch operations. Works
+    recursively, so a common chain of patches will be resolved as well.
+    """
+    assert isinstance(path, (tuple, list))
+    popped = _pop_path(diffs)
+    while popped:
+        path = path + (popped["key"],)
+        diffs = popped["diffs"]
+        popped = _pop_path(diffs)
+    return path, diffs
+
+
+def _pop_path(diffs):
+    """Pops of a common path from patch ops sharing a common key.
+
+    Checks whether all diffs are single patch operations sharing the same key,
+    or alternatively empty diffs. If so, it returns the shared path/key as well
+    as the inner diffs of the patch operations (in the same order as the passed
+    diffs).
+    """
+    key = None
+    popped_diffs = []
+    for d in diffs:
+        # Empty diffs can be skipped
+        if d is None or len(d) == 0:
+            popped_diffs.append(None)
+            continue
+        # Check that we have only one op, which is a patch op
+        if len(d) != 1 or d[0].op != DiffOp.PATCH:
+            return
+        # Ensure all present diffs have the same key
+        if key is None:
+            key = d[0].key
+        elif key != d[0].key:
+            return
+        # Ensure the sub diffs of all ops are suitable as outer layer
+        # if d[0].diff.length > 1:
+        #    return
+        popped_diffs.append(d[0].diff)
+    if key is None:
+        return
+    return {'key': key, 'diffs': popped_diffs}
+
+
+def push_path(path, diffs):
+    """Wraps the diffs in patch operations matching path.
+    """
+    for key in reversed(path):
+        diffs = [op_patch(key, diffs)]
+    return diffs
+
+
+def pop_patch_decision(decision):
+    """Create a new decision one level lower in diff tree.
+
+    Checks whether all diffs are single patch operations sharing the same key,
+    or alternatively empty diffs. A new decision is then created at that key.
+
+    Returns the new decision.
+
+    Raises a ValueErorr if a decision can not be created at the lower level.
+    """
+    diffs = [decision.local_diff, decision.remote_diff]
+    if decision.action == "custom":
+        diffs.append(decision.custom_diff)
+    popped = _pop_path(diffs)
+    if popped is None:
+        raise ValueError("Cannot pop patch decision for: " + str(decision))
```

Comment:
Seems like you're using exceptions for expected program flow, may I suggest `return None` instead? (see below)